### PR TITLE
(Improvement) Request Model

### PR DESCRIPTION
**Pull request by 99Anvar99**
- We don't want to request models that does not exist! or invalid

### CODE REVIEW

File: src/util/entity.cpp [hunk 1]
```diff
@@ -218,7 +219,15 @@ namespace big::entity
 	        return false;
 	    }
 
-	    location.z = ground_z + 1.f;
+        if (WATER::GET_WATER_HEIGHT(location.x, location.y, location.z, &height))
+        {
+			location.z = height;
+        }
+		else
```
Comment (Yimura) [inline src/util/entity.cpp]: Mixed indentation, run `clang-format -i src/**.hpp src/**.cpp` to fix.

File: src/util/ped.hpp [hunk 2]
```diff
@@ -446,17 +446,20 @@ namespace big::ped
...
-		for (uint8_t i = 0; !STREAMING::HAS_MODEL_LOADED(hash) && i < 100; i++)
+		if (STREAMING::IS_MODEL_VALID(hash) && STREAMING::IS_MODEL_IN_CDIMAGE(hash) && !STREAMING::HAS_MODEL_LOADED(hash))
 		{
-			STREAMING::REQUEST_MODEL(hash);
-			script::get_current()->yield();
+			while (!STREAMING::HAS_MODEL_LOADED(hash))
+			{
+				STREAMING::REQUEST_MODEL(hash);
+				script::get_current()->yield();
+			}
```
Comment (Yimura) [inline src/util/ped.hpp]: Since this piece of code is being duplicated quite a bit, how about making it a general purpose helper function?

**Review by Yimura — CHANGES_REQUESTED** [review]
Just some mixed indentations here and there as well as duplication of commonly used code.

### To Do:
 - [x] Move the code block that checks if a model is valid and then loads it to a helper function
 - [x] Run clang-format to make the code that you submit uniform with the main code base.

File: src/util/ped.hpp (hunk 2 above)
Comment (99Anvar99) [reply src/util/ped.hpp]: Sure thing.

**Comment by 99Anvar99** [comment]
My bad, i meant to say created helper for request model function not class.

Edit: Everything seems to work fine!

File: src/util/entity.cpp [hunk 3]
```diff
@@ -193,33 +193,67 @@ namespace big::entity
...
+
+		if (STREAMING::IS_MODEL_VALID(hash) && STREAMING::IS_MODEL_IN_CDIMAGE(hash) && !has_loaded)
+		{
+			while (!has_loaded)
+			{
+				if (has_loaded)
+					break;
+				STREAMING::REQUEST_MODEL(hash);
+				script::get_current()->yield();
+			}
```
Comment (Yimura) [inline src/util/entity.cpp]: `has_loaded` is never updated in this loop, might cause a deadlock.

File: src/util/entity.cpp [hunk 4]
```diff
@@ -193,33 +193,67 @@ namespace big::entity
...
+			return true;
+		}
+
+		if (has_loaded)
+		{
+			return true;
+		}
+
+		return false;
```
Comment (Yimura) [inline src/util/entity.cpp]: Can be changed to `return has_loaded`.

File: src/util/world_model.hpp [hunk 5]
```diff
@@ -15,26 +15,21 @@ namespace big::world_model
...
-		}
+			std::call_once(once_flag, setup_backup);
+			memset(g_pointers->m_gta.m_world_model_spawn_bypass, 0x90, patch_size);
 
-		std::call_once(once_flag, setup_backup);
-		memset(g_pointers->m_gta.m_world_model_spawn_bypass, 0x90, patch_size);
+			const auto object = OBJECT::CREATE_OBJECT(hash, location.x, location.y, location.z, is_networked, false, false);
 
-		const auto object = OBJECT::CREATE_OBJECT(hash, location.x, location.y, location.z, is_networked, false, false);
+			memcpy(g_pointers->m_gta.m_world_model_spawn_bypass, backup.data(), patch_size);
```
Comment (Yimura) [inline src/util/world_model.hpp]: I know this isn't your fault but could you update the patching of the `m_world_model_spawn_bypass` and make it into a `memory::byte_patch`?

Similarly to how it was implemented here:
https://github.com/YimMenu/YimMenu/blob/7efbee2e85ba2aee773e398cd9e66d54e640aca5/src/byte_patch_manager.cpp#L16

And here's an example usage:
https://github.com/YimMenu/YimMenu/blob/7efbee2e85ba2aee773e398cd9e66d54e640aca5/src/backend/looped/self/police.cpp#L19

**Review by Yimura — CHANGES_REQUESTED** [review]

File: src/util/world_model.hpp (hunk 5 above)
Comment (99Anvar99) [reply src/util/world_model.hpp]: Yes, i will update it.

**Review by Yimura — APPROVED** [review]
Thanks for another PR fixing and improving some of the tech debt in YimMenu!

**Pull request merged**